content/en/agent/guide/python-3.md

```diff
@@ -0,0 +1,265 @@
+```
+
+### Six
+
+[Six][5] is a Python 2/3 compatibility library intended to allow developers to ship ubiquitous Python code. Some of the examples below make use of six to make legacy Python 2 code compatible with Python 3.
+
+### Dictionary methods
+
+In Python 3, the `dict.iterkeys()`, `dict.iteritems()` and `dict.itervalues()` methods are not available.
+
+| Python 2 | Python 2 and 3 |
+| --- | --- |
+| `for key in mydict.iterkeys():` <br/> &nbsp;&nbsp;`  ...` | `for key in mydict:`<br/> &nbsp;&nbsp;`  ...` |
+| `for key, value in mydict.iteritems():`<br/> &nbsp;&nbsp;`  ...` | `from six import iteritems` <br/><br/> `for key, value in iteritems(mydict):`<br/> &nbsp;&nbsp;`  ...`|
+| `for value in mydict.itervalues():`<br/> &nbsp;&nbsp;`  ...` | `from six import itervalues` <br/><br/> `for value in itervalues(mydict):`<br/> &nbsp;&nbsp;`  ...` |
+
+Also, in Python 3, the `dict.keys()`, `dict.items()`, `dict.values()` methods return iterators. Therefore, if the dictionary needs to be modified during iteration, it's necessary to make a copy first. To retrieve a dictionary’s keys/items/values as a list:
+
+| Python 2 | Python 2 and 3 |
+| --- | --- |
+| `mykeylist = mydict.keys()` | `mykeylist = list(mydict)` |
+| `myitemlist = mydict.items()` | `myitemlist = list(mydict.items())` |
+| `myvaluelist = mydict.values()` | `myvaluelist = list(mydict.values()` |
+
+The `dict.has_key()` method is deprecated in Python 2 and is removed in Python 3. Use the `in` operator instead.
+
+| Python 2 | Python 2 and 3 |
+| --- | --- |
+| `mydict.has_key('foo') //deprecated` | `foo in mydict` |
+
+### Standard Library Changes
+
+Python 3 features a reorganized standard library, where a number of modules and functions were renamed or moved. Importing moved modules through `six.moves` works on both Python versions.
+
+| Python 2 | Python 3 | Python 2 and 3 |
+| --- | --- | --- |
+| `import HTMLParser` | `import html.parser` | `from six.moves import html_parser` |
+
+Consult the [six documentation][6] for the list of renamed modules. Note that the `urllib`, `urllib2`, and `urlparse` modules have been heavily reorganized.
+
+### Unicode
+
+Python 2 treats unicode text and binary-encoded data in the same manner, and tries to automatically convert between bytes and strings as needed. This works as long as all characters are ASCII, but leads to unexpected behavior when non-ASCII characters are encountered.
+
+| type | literal | Python 2 | Python 3 |
+| --- | --- | --- | --- |
+| bytes | b'...' | binary | binary |
+| str | '...' | binary | text |
+| unicode | u'...' | text | text |
+
+Text data are Unicode code points; you must encode with `.encode(encoding)` for storage or transmission. Binary data are encoded code points represented as a sequence of bytes that must be decoded with `.decode(encoding)` back to text. When reading text from a file, the `open` function from the `io` package is handy because the data read is already decoded into Unicode:
```

Comment:
```suggestion
Text data is Unicode code points; you must encode with `.encode(encoding)` for storage or transmission. Binary data is encoded code points represented as a sequence of bytes that must be decoded with `.decode(encoding)` back to text. When reading text from a file, the `open` function from the `io` package is handy because the data read is already decoded into Unicode:
```